### PR TITLE
added role based authorization directive

### LINKED ISSUE #298: [Feature request] Custom directive for role based authorization
**Is your feature request related to a problem? Please describe.**
Currently, the role-based authorization is being handled by a helper function at the resolver level. This makes for reduced flexibility in terms of how we can restrict access.

**Describe the solution you'd like**
We can make use of a custom directive that handles role-based authorization at the schema level. This provides us with more flexibility in terms of restricting access. Be it restricting a complete query or just a field. In addition, it separates out the authorization logic from the business logic.

**Describe alternatives you've considered**
None

**Approach to be followed (optional)**
Creation of a custom authentication directive and handling authorization at the schema level.

**Additional context**
I would like to work on this under the guidance of @sumitra19jha

### PR DESCRIPTION
**Pull request by aniketbiswas21**
**What kind of change does this PR introduce?**

Added a role-based authorization directive that checks if the requesting user is a user/admin/super-admin

**Issue Number:**

Fixes #298 

**Did you add tests for your changes?**

No

**Snapshots/Videos:**

https://user-images.githubusercontent.com/51146347/125109108-8c001f80-e100-11eb-8c76-746cde4fe068.mp4

**If relevant, did you update the documentation?**

Not required

**Does this PR introduce a breaking change?**

No

**Other information**

N/A

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by aniketbiswas21** [comment]
@sumitra19jha @palisadoes I have added the directive. Please check!

Also, Can you tell me in what all query/mutation I have to use this?

**Comment by sumitra19jha** [comment]
Can you please share the snapshot of the demo with one query or mutation?

**Comment by aniketbiswas21** [comment]
> Can you please share the snapshot of the demo with one query or mutation?

In which query should I demo it with?

**Comment by sumitra19jha** [comment]
You can use any query for the demo purpose but don't push a commit on that

**Comment by aniketbiswas21** [comment]
> You can use any query for the demo purpose but don't push a commit on that

Added a video. Please check

File: directives/roleDirective.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+const requestContext = require('talawa-request-context');
+const userExists = require('../helper_functions/userExists');
+
+class RoleAuthorizationDirective extends SchemaDirectiveVisitor {
+  visitFieldDefinition(field) {
+    const resolver = field.resolve || defaultFieldResolver;
+    const { requires } = this.args;
+    field.resolve = async (root, args, context, info) => {
+      const user = await userExists(context.userId);
+      console.log(user.userType);
```
Comment (sumitra19jha) [inline directives/roleDirective.js:13]: Remove the console statement

File: schema/schema.graphql [hunk 2]
```diff
@@ -11,7 +11,14 @@ const utils = require("./utils.graphql");
 
 const typeDefs = gql`
   directive @auth on FIELD_DEFINITION
+  directive @role(requires: Role) on FIELD_DEFINITION
 
+  enum Role {
```
Comment (sumitra19jha) [inline schema/schema.graphql:16]: The user.graphql file already contains the `UserType` enum. Please remove this duplicate code and use that

**Review by sumitra19jha — COMMENTED** [review]
The PR is good to be merge. Just update these two changes

File: directives/roleDirective.js (hunk 1 above)
Comment (aniketbiswas21) [reply directives/roleDirective.js]: Fixed

File: schema/schema.graphql (hunk 2 above)
Comment (aniketbiswas21) [reply schema/schema.graphql]: Fixed

**Comment by aniketbiswas21** [comment]
@sumitra19jha I have made the required changes. Please take a look!

**Review by sumitra19jha — APPROVED** [review]

**Comment by palisadoes** [comment]
@DangaRanga @ranilwallace please review

**Comment by aniketbiswas21** [comment]
@palisadoes @DangaRanga @ranilwallace Sir, is the PR fine? :sweat_smile:

**Comment by palisadoes** [comment]
@DangaRanga @ranilwallace @xsh2047 Please review

**Review by Sagar2366 — APPROVED** [review]

**Pull request merged**